### DISCUSSION TITLE
Multiple USB ports on the SAMD21?

### DISCUSSION THREAD

**Discussion by shraiwi** [hathach/tinyusb#1135] [Q&A] [answered]
Hi, is it possible to use the SAMD21's built-in USB interface along with this library to provide one hardware and one bit-banged USB interface? I'm trying to connect two computers, with one using USB serial (in hardware) and the other emulating an HID keyboard+mouse combo (using TinyUSB, if possible). Each will be on separate physical USB Port. Thank you!

Edit: I don't really care about which USB interface is in hardware and which one's in software, I'm just wondering if it's even possible to add an additional one on top of the hardware one using TinyUSB

**Comment by hathach** [comment, marked as answer]
multiple controllers used to be supported but dropped due to the complexity of API with additional controller arguments and virtually un-used in practical. You could try to get it work yourself, it is not very complicated changes.

**Reply by shraiwi** [reply]
Thanks for the answer. Is there any documentation how to do it myself? If not, what changes are required?

**Reply by hathach** [reply]
- no document for this purpose
- you need to change pretty much most of the stack API, driver and internal functions. However, it is a systematic change, if you got familiar enough with the stack. It is not too difficult to do.

**Reply by shraiwi** [reply]
Alright, I'll look into it. Thank you!
